### PR TITLE
Add tofunft trade scraper (Astar network)

### PR DESCRIPTION
**Pull request by hhio618**
### Description
Add TofuNFT trade scraper for Astar network

### CODE REVIEW

File: pkg/dia/nft/nftTrade-scrapers/tofunft.go [hunk 1]
```diff
@@ -0,0 +1,772 @@
...
+	if err != nil {
+		log.Errorf("unable to read transaction(%s) receipt: %s", tx.TXHash, err.Error())
+		return false, err
+	}
+
+	currSymbol := "ASTR"
+	currAddr := common.Address{}
+	currDecimals := 18
+
+	transfers, err := s.findERC721Transfers(ctx, receipt)
```
Comment (jppade) [inline pkg/dia/nft/nftTrade-scrapers/tofunft.go:381]: Please review this method / its applicability here. The resulting transfer object does not hold the correct information. I.e.
- transfer.From is always zero address
- transfer.To is the NG Contract address
- transfer.NFTAddress is WASTR token address
- transfer.tokenID seems to be the price (paid in WASTR)

File: pkg/dia/nft/nftTrade-scrapers/tofunft.go [hunk 2]
```diff
@@ -0,0 +1,772 @@
...
+		return err
+	}
+
+	nft, err := s.createOrReadNFT(nftClass, transfer)
+	if err != nil {
+		return err
+	}
+
+	// Get block time.
+	timestamp, err := ethhelper.GetBlockTimeEth(int64(ev.Raw.BlockNumber), s.tradeScraper.datastore, s.tradeScraper.ethConnection)
```
Comment (jppade) [inline pkg/dia/nft/nftTrade-scrapers/tofunft.go:430]: Please fetch the timestamp using s.tradeScraper.ethConnection.BlockByNumber(...) as this method is only applicable for Eth mainnet.

Comment (jppade) [inline pkg/dia/nft/nftTrade-scrapers/tofunft.go:430]: Please fetch the timestamp using s.tradeScraper.ethConnection.BlockByNumber(...) as this method is only applicable for Eth mainnet.

File: pkg/dia/nft/nftTrade-scrapers/tofunft.go [hunk 3]
```diff
@@ -0,0 +1,772 @@
...
+	log.Infof("processed %d trades", numTrades)
+
+	return nil
+}
+
+func (s *TofuNFTScraper) processTx(ctx context.Context, tx *utils.EthFilteredTx) (bool, error) {
+	log.Tracef("process tx -> block: %d, tx index: %d, tx hash: %s", tx.BlockNum, tx.TXIndex, tx.TXHash.Hex())
+
+	var ev tofunft.TofunftEvInventoryUpdate
+	err := tofuNFTABI.UnpackIntoInterface(&ev, "EvInventoryUpdate", tx.Logs[0].Data)
```
Comment (jppade) [inline pkg/dia/nft/nftTrade-scrapers/tofunft.go:365]: The resulting 0object ev does not hold the data as used later on:
- ev.Raw.BlockNumber is always zero
- ev.Raw.TxHash is always zero

Comment (jppade) [inline pkg/dia/nft/nftTrade-scrapers/tofunft.go:365]: The resulting 0object ev does not hold the data as used later on:
- ev.Raw.BlockNumber is always zero
- ev.Raw.TxHash is always zero

**Review by jppade — COMMENTED** [review]
Many thanks for the pull request. There is still a couple of issues regarding the data quality.

**Comment by hhio618** [comment]
@jppade Thanks for the review! I fixed the above problems. also, this is the command I'm using to test the scraper:
`ETH_URI_REST=https://evm.astar.network go run cmd/nftTradescrapers/main.go -nftclass TofuNFT`

**Pull request merged**